### PR TITLE
[ELF] Support relocatable files using CREL with explicit addends

### PR DESCRIPTION
**Pull request by MaskRay**
... using the temporary section type code 0x40000020
(`clang -c -Wa,--crel,--allow-experimental-crel`). LLVM will change the
code and break compatibility (Clang and lld of different versions are
not guaranteed to cooperate, unlike other features). CREL with implicit
addends are not supported.

---

Introduce `RelsOrRelas::crels` to iterate over SHT_CREL sections and
update users to check `crels`.

(The decoding performance is critical and error checking is difficult.
Follow `skipLeb` and `R_*LEB128` handling, do not use
`llvm::decodeULEB128`, whichs compiles to a lot of code.)

A few users (e.g. .eh_frame, LLDDwarfObj, s390x) require random access. Pass
`/*supportsCrel=*/false` to `relsOrRelas` to allocate a buffer and
convert CREL to RELA (`relas` instead of `crels` will be used). Since
allocating a buffer increases, the conversion is only performed when
absolutely necessary.

---

Non-alloc SHT_CREL sections may be created in -r and --emit-relocs
links. SHT_CREL and SHT_RELA components need reencoding since
r_offset/r_symidx/r_type/r_addend may change. (r_type may change because
relocations referencing a symbol in a discarded section are converted to
`R_*_NONE`).

* SHT_CREL components: decode with `RelsOrRelas` and re-encode (`OutputSection::finalizeNonAllocCrel`)
* SHT_RELA components: convert to CREL (`relToCrel`). An output section can only have one relocation section.
* SHT_REL components: print an error for now.

SHT_REL to SHT_CREL conversion for -r/--emit-relocs is complex and
unsupported yet.

Link: https://discourse.llvm.org/t/rfc-crel-a-compact-relocation-format-for-elf/77600

### CODE REVIEW

**Comment by smithp35** [comment]
Will take a look as soon as I can. I'll be at a conference this week so may be a bit slow.

**Comment by smithp35** [comment]
> Will take a look as soon as I can. I'll be at a conference this week so may be a bit slow.

I've started looking, there's quite a bit to go through. I don't have any comments right now, still tracing through the changes. Hope to have something by the end of the week.

**Comment by DEBADRIBASAK** [comment]
> > Will take a look as soon as I can. I'll be at a conference this week so may be a bit slow.
> 
> I've started looking, there's quite a bit to go through. I don't have any comments right now, still tracing through the changes. Hope to have something by the end of the week.

Thank you so much for looking into this. This PR is the remaining part of the CREL launch and one of our internal action items is blocked on this. So, I will be highly grateful if we get an approval on this.
Thank you in advance.

File: lld/ELF/ICF.cpp
Comment (arichardson) [inline lld/ELF/ICF.cpp:327]: Should this also have `|| rb.areRelocsCrel()` like the rel case below?

File: lld/ELF/ICF.cpp
Comment (arichardson) [inline lld/ELF/ICF.cpp:379]: Sam's question here.

File: lld/ELF/InputSection.cpp
Comment (arichardson) [inline lld/ELF/InputSection.cpp:164]: Could use llvm::enumerate() here?

**Review by arichardson — COMMENTED** [review]
Did not manage to review all the code, just some small comments from the first few files

File: lld/ELF/ICF.cpp
Comment (MaskRay) [reply lld/ELF/ICF.cpp:327]: Not needed. The double rel check is for 686cff17cc310884e48ae963bf7507f96950cc90 .

An early `areRelocsCrel()` issue would not run into that issue.

File: lld/ELF/ICF.cpp
Comment (MaskRay) [reply lld/ELF/ICF.cpp:379]: Not needed. The double rel check is for https://github.com/llvm/llvm-project/commit/686cff17cc310884e48ae963bf7507f96950cc90 .

An early areRelocsCrel() issue would not run into that issue.

File: lld/ELF/InputSection.cpp
Comment (MaskRay) [reply lld/ELF/InputSection.cpp]: Thanks for the suggestion. Added `iterator_category/value_type/etc` boilerplate to support `llvm::enumerate`.

**Comment by smithp35** [comment]
My apologies for going quiet. I'm hoping to spend some time tomorrow going through this.

**Review by smithp35 — COMMENTED** [review]
I've made one pass through all the code. I've scanned the tests but haven't gone into much detail yet to see if there's anything missing.

File: lld/ELF/Relocations.h
Comment (MaskRay) [reply lld/ELF/Relocations.h:271]: The commend `// R_PPC64_REL24_NOTOC in Relocations.cpp` can be expanded. The `operator*()` is used by 
` end = static_cast<const void *>(rels.end()); ` in Relocations.cpp

Comment (MaskRay) [reply lld/ELF/Relocations.h]: `end` is for additional error checking for PPC64 TLS. The error checking is technically not that useful. I removed the `operator` here and reworked the ppc64 special case.

**Review by MaskRay — COMMENTED** [review]
.

**Comment by MaskRay** [comment]
There is still time to include this patch before  the LLVM 19.1-rc2 release on August 6th.
However, introducing new features after that could be challenging.
The 19.1 release will showcase the impressive relocatable size reduction to the wider ELF OS community.

**Comment by smithp35** [comment]
My apologies for the delay. I'm intending to go through the tests in more detail tomorrow. I don't have any outstanding comments on the implementation.

**Comment by smithp35** [comment]
I've taken a look at the tests. I'm now down to a few last questions, triggered by the no mixing of SHT_REL and CREL in emit-relocs and -r links.

I was trying to work out why it would be complex. As I understand it:
* CREL in relocatable objects uses explicit addends even when it is usual to use SHT_REL. (From Addend bit: in https://maskray.me/blog/2024-03-09-a-compact-relocation-format-for-elf)
* To convert from SHT_REL to CREL would require extracting the implicit addends

**Comment by MaskRay** [comment]
> I've taken a look at the tests. I'm now down to a few last questions, triggered by the no mixing of SHT_REL and CREL in emit-relocs and -r links. I can probably figure this out by running, but I'm hoping I can speed up the process by asking.
> 
> I was trying to work out why it would be complex. As I understand it:
>
> * CREL in relocatable objects uses explicit addends even when it is usual to use SHT_REL. (From Addend bit: in https://maskray.me/blog/2024-03-09-a-compact-relocation-format-for-elf)

Yes, the assembler implementation is currently kept simple and assumes the explicit addend form.
Therefore, lld doesn't have to deal with implicit addend input.

> * To convert from SHT_REL to CREL would require extracting the implicit addends from the REL relocations

Yes. We would need `target.getImplicitAddend`.

> * In theory this addend should be encodeable in CREL.

Since multiple relocation sections aren't supported, when combining input sections, their relocation sections, if mixing CREL/RELA, have to be combined as well.
CREL is the desired output format for size.

> * When applying the relocation we'll need to be careful that the implict addend is overwritten, I think this should always be the case for SHT_REL.

Yes.

> Is there anything I've missed?

No:)

> Following up. If CREL in relocatable object files is supposed to use explicit addends, how do we deal with an input that uses implicit addends, or at least a non 0 implicit addend? At a glance it looks like we might assume that there is always an explicit addend, but I don't know the code well enough to easily work that out. I'll try again tomorrow.

There is currently no error checking. 
A diagnostic could be added by changing `ret.crels = Relocs<typename ELFT::Crel>` in `InputSectionBase::relsOrRelas` to add an extra argument to indicate the section name.
There will be some minor overhead.

File: lld/ELF/MarkLive.cpp
Comment (smithp35) [inline lld/ELF/MarkLive.cpp:88]: Can we add a comment that we assume all input Crel relocations will have an explicit addend?

**Review by smithp35 — COMMENTED** [review]
Thanks for answering the questions.

I think we will need to support `SHT_REL` to `CREL` conversion for relocatable links and emit-relocs will likely not be supported with `CREL` on `SHT_REL` platforms. I'm thinking that almost all links will have some files from archives like `crt0.o` that will have `SHT_REL` relocations. 

I also think we'll need to have some kind of way of either requiring input CREL relocations to have explicit addends (with diagnostic), or we support implicit addends.

I'm fine with these being limitations for an experimental CREL implementation though. We can fix these in further patches.

My last suggestion. Can we add an entry into the release notes about supporting CREL and either link or describe what the limitations are?

I'm happy to Approve after that.

**Comment by MaskRay** [comment]
> Thanks for answering the questions.
> 
> I think we will need to support `SHT_REL` to `CREL` conversion for relocatable links and emit-relocs will likely not be supported with `CREL` on `SHT_REL` platforms. I'm thinking that almost all links will have some files from archives like `crt0.o` that will have `SHT_REL` relocations.
> 
> I also think we'll need to have some kind of way of either requiring input CREL relocations to have explicit addends (with diagnostic), or we support implicit addends.
> 
> I'm fine with these being limitations for an experimental CREL implementation though. We can fix these in further patches.
> 
> My last suggestion. Can we add an entry into the release notes about supporting CREL and either link or describe what the limitations are?
> 
> I'm happy to Approve after that.

Thanks! I intentionally omit the `lld/docs/ReleaseNotes.rst` change to avoid back-port complexity. The release note will be added separately. How about this wording?

```
Experimental CREL relocations with explicit addends are now supported using the
temporary section type code 0x40000020 (`clang -c -Wa,--crel,--allow-experimental-crel`).
LLVM will change the code and break compatibility. CREL with implicit addends
are not supported.
```

**Review by smithp35 — APPROVED** [review]
Yes, that's fine we can add the release notes in a follow up patch.

**Comment by MaskRay** [comment]
> Yes, that's fine we can add the release notes in a follow up patch.

Thanks so much for the extensive suggestions. Will do:)

**Pull request merged**
